### PR TITLE
Make flutter tool enforce >= kotlin 1.7.0, Gradle 7.0.2, and AGP 7.0.0, and Fix test failures blocking androidx upgrade

### PR DESCRIPTION
**Pull request by gmackall**
PR to pave the way for https://github.com/flutter/engine/pull/53001 to re-land

Summary: 
- Enforces use of Kotlin >= `1.7.0` (please see below note)
- Fixes ci failures that prevented the above PR from landing.

Details:

Because it landed initially, we are able to fake the roll in this PR to fix all the tests ([see my comment](https://github.com/flutter/flutter/pull/149204#discussion_r1617924772)).

Fixes all the tests that failed:
1. `module_test` failing on multiple platforms (3/9 of the failures). 
Failure is 
```
> Android resource linking failed
   ERROR:/b/s/w/ir/x/t/flutter_module_test.KECMXW/hello/.android/plugins_build_output/device_info/intermediates/merged_res/release/values/values.xml:194: AAPT: error: resource android:attr/lStar not found.
```

This is a rather unhelpful error message but some [folks online suggest](https://stackoverflow.com/a/69050529) that upgrading your `compileSdk` version fixes this.
These resolve when I remove the dependency on the long discontinued [package_info](https://pub.dev/packages/package_info) and [device_info](https://pub.dev/packages/device_info) packages, perhaps because they are transitively pulling in low `compileSdk` versions? This is unclear to me.

2. `module_custom_host_app_name_test` was failing for the same reason (another 3/9, or cumulative 6/9).
3. `tool_integration_tests_3_4` was a flake 🙂 (7/9)
4. `framework_tests_slow` needed a newer version of the Kotlin Gradle plugin (the flutter tool tells us this, so I just upgraded as suggested) and it resolved (8/9)
5.`android_preview_tool_integration_tests` needed newer AGP and KGP versions. I also refactored the tests, and bumped our error versions, fixing https://github.com/flutter/flutter/issues/142653.

**Note that the bump to KGP is not in line with our policy** - we didn't warn for `1.5.0-1.6.x` for a release (or at all) before dropping support. But I think it might still be justified:
- The bump to our androidx libraries unblocks ongoing Scribe work, and also includes a fix for a [memory leak](https://github.com/flutter/flutter/issues/129307#issuecomment-1601636959) and a [crash on folding phones](https://github.com/flutter/flutter/issues/114868#issuecomment-2133226962), among many other bug fixes.
- Gradle [doesn't test on half of that range](https://docs.gradle.org/current/userguide/compatibility.html#kotlin), and so we implicitly can't claim to support it either. More generally, our Java and Kotlin support ranges should probably strictly fall within what Gradle tests.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

**Comment by gmackall** [comment]
1. `module_test` failing on multiple platforms (3/9 of the failures). 
Failure is 
```
> Android resource linking failed
   ERROR:/b/s/w/ir/x/t/flutter_module_test.KECMXW/hello/.android/plugins_build_output/device_info/intermediates/merged_res/release/values/values.xml:194: AAPT: error: resource android:attr/lStar not found.
```

This is a rather unhelpful error message but some [folks online suggest](https://stackoverflow.com/a/69050529) that upgrading your `compileSdk` version fixes this.
These resolve when I remove the dependency on the long discontinued [package_info](https://pub.dev/packages/package_info) and [device_info](https://pub.dev/packages/device_info) packages, perhaps because they are transitively pulling in low `compileSdk` versions? This is unclear to me.

File: bin/internal/engine.version [hunk 1]
```diff
@@ -1 +1 @@
-d0323905fc2f62019d10110ecc0b3974e8df3ca7
+00e72e9cecb138e001f5625ff67a370aabd3e3c5
```
Comment (gmackall) [inline bin/internal/engine.version:1]: This should NOT be landed. It just points this PR at my [earlier engine PR](https://github.com/flutter/engine/pull/53001), for the purposes of getting tests passing.

**Comment by gmackall** [comment]
Three tests fail in `android_preview_tool_integration_tests` - two of the failures are fixed by upgrading AGP versions, which we can technically do according to the policy at https://github.com/flutter/flutter/issues/140913. But the last one requires a newer version of Kotlin (at least 1.7.10), which would break the policy currently unfortunately

**Comment by gmackall** [comment]
Bonus, the refactor to the dependency version checker tests should make it very easy for us to set up a suite of tests that range our Gradle/AGP/KGP versions

**Review by reidbaker — COMMENTED** [review]
I cant tell if this is a staging pr or one that you intend to land. The engine.version is still present that your comment says should not be landed but this is not a draft and this pr set the minimum kotlin version. 

If this is the pr to set the error version of kotlin then lets me that the title and first points in the pr description.

**Comment by reidbaker** [comment]
@stuartmorgan adding you as the voice of why we should not immediately error on versions of kotlin older than 1.7.

**Review by johnmccutchan — APPROVED** [review]
> The bump to our androidx libraries unblocks ongoing Scribe work, and also includes a fix for a https://github.com/flutter/flutter/issues/129307#issuecomment-1601636959, among many other bug fixes.

The Scribe work is important to our collaboration with Android folks. We need to land this.

**Comment by reidbaker** [comment]
> > @stuartmorgan adding you as the voice of why we should not immediately error on versions of kotlin older than 1.7.
> 
> This is purely an application-level change, right? I.e., if a developer hits this error, they just need to change a setting in their local code, rather than needing to get the authors of all plugins they use to update?
> 
> If that's correct, I don't have any objection; while it's a little bumpier for developers without a warning as compared to with one, I'm significantly less concerned about breakage that individual developers can resolve entirely locally than things that need to flow upstream from broken developers to plugin authors and then though the ecosystem and back to developers, as the latter is (from what I've seen) orders of magnitude more disruptive.

Thank you stuart for being the voice of the customer.

File: packages/flutter_tools/test/android_preview_integration.shard/android_dependency_version_checking_test.dart [hunk 2]
```diff
@@ -108,101 +107,39 @@ void main() {
...
     ], workingDirectory: app.path);
-    expect(result, const ProcessResultMatcher());
-    expect(result.stderr, contains('Please upgrade your Android Gradle '
-        'Plugin version'));
-  });
+    return result;
+  }
 
   testUsingContext(
-      'Gradle version out of "warn" support band prints warning but still builds', () async {
```
Comment (reidbaker) [inline packages/flutter_tools/test/android_preview_integration.shard/android_dependency_version_checking_test.dart:117]: I think we still need this test even if it means we need to build a way to override the warn/error version so there can be a version that is in warn but not error.

**Review by reidbaker — CHANGES_REQUESTED** [review]

**Comment by gmackall** [comment]
> > @stuartmorgan adding you as the voice of why we should not immediately error on versions of kotlin older than 1.7.
> 
> This is purely an application-level change, right? I.e., if a developer hits this error, they just need to change a setting in their local code, rather than needing to get the authors of all plugins they use to update?
> 
> If that's correct, I don't have any objection; while it's a little bumpier for developers without a warning as compared to with one, I'm significantly less concerned about breakage that individual developers can resolve entirely locally than things that need to flow upstream from broken developers to plugin authors and then though the ecosystem and back to developers, as the latter is (from what I've seen) orders of magnitude more disruptive.

Yes, that is correct. That is true for both this tool change, and for the error that users might encounter as a result of the re-land of https://github.com/flutter/engine/pull/53001

**Comment by reidbaker** [comment]
We talked offline and there are some difficulties in testing something that is run in gradle and needs gradle objects but is actually a script. For now the compromise was to include a patch version delta so that we could ensure the warning logic does not break.

**Comment by gmackall** [comment]
> We talked offline and there are some difficulties in testing something that is run in gradle and needs gradle objects but is actually a script. For now the compromise was to include a patch version delta so that we could ensure the warning logic does not break.

It isn't possible to bump a Gradle patch version and test that warn prints (I'd have to bump a minor version, because our minimum supported AGP version already requires the maximum patch version of AGP `7.0.x`).

So I've left that test out, but included tests that we print warnings and don't block the build for AGP and KGP. Let me know if you'd prefer I bump the "warn" minor version for Gradle to `7.1.0` so we can include that test as well.

File: packages/flutter_tools/test/android_preview_integration.shard/android_dependency_version_checking_test.dart (hunk 2 above)
Comment (gmackall) [reply packages/flutter_tools/test/android_preview_integration.shard/android_dependency_version_checking_test.dart:117]: I've bumped the patch versions, and restored the tests. Let me know if it still looks good!

**Review by reidbaker — APPROVED** [review]

**Review by reidbaker — APPROVED** [review]

**Comment by gmackall** [comment]
Mac builders seem to be timing out very frequently right now

**Pull request merged**